File: market-data/crypto-news/examples/x402-client.py
#!/usr/bin/env python3
"""
x402 Payment Client Example - Python

This example shows how to make authenticated API calls using x402 payments.

Requirements:
    pip install requests eth-account web3

Usage:
    export WALLET_PRIVATE_KEY="your_private_key_here"
    python x402-client.py

See: https://docs.x402.org
"""

import os
import json
import time
import hashlib
from typing import Optional, Dict, Any
from dataclasses import dataclass

import requests
from eth_account import Account
from eth_account.messages import encode_defunct
from web3 import Web3

# =============================================================================
# CONFIGURATION
# =============================================================================

API_BASE = "https://cryptocurrency.cv"
PRIVATE_KEY = os.environ.get("WALLET_PRIVATE_KEY")

if not PRIVATE_KEY:
    print("Warning: WALLET_PRIVATE_KEY not set. Using demo mode.")
    PRIVATE_KEY = "0x" + "0" * 64  # Dummy key for demo

# Create account from private key
account = Account.from_key(PRIVATE_KEY)
print(f"Wallet address: {account.address}")


# =============================================================================
# X402 PAYMENT HANDLING
# =============================================================================

@dataclass
class PaymentRequirements:
    """Parsed 402 Payment Required response"""
    version: int
    network: str
    amount: str
    pay_to: str
    asset: str
    nonce: str
    resource: str


def parse_402_response(data: Dict[str, Any]) -> Optional[PaymentRequirements]:
    """Parse x402 payment requirements from 402 response"""
    x402 = data.get("x402", {})
    accepts = x402.get("accepts", [])
    
    if not accepts:
        return None
    
    # Use first accepted payment method
    accept = accepts[0]
    
    return PaymentRequirements(
        version=x402.get("x402Version", 1),
        network=accept.get("network", ""),
        amount=accept.get("maxAmountRequired", "0"),
        pay_to=accept.get("payTo", ""),
        asset=accept.get("asset", ""),
        nonce=accept.get("paymentNonce", ""),
        resource=accept.get("resource", ""),
    )


def create_payment_signature(requirements: PaymentRequirements) -> str:
    """
    Create x402 payment signature
    
    Note: This is a simplified example. Production implementations
    should use the official x402 SDK which handles the full signing flow.
    """
    # Create payment message
    message = json.dumps({
        "network": requirements.network,
        "amount": requirements.amount,
        "payTo": requirements.pay_to,
        "asset": requirements.asset,
        "nonce": requirements.nonce,
        "resource": requirements.resource,
        "timestamp": int(time.time()),
    }, separators=(',', ':'))
    
    # Sign with wallet
    message_hash = encode_defunct(text=message)
    signature = account.sign_message(message_hash)
    
    return signature.signature.hex()


def make_paid_request(endpoint: str, method: str = "GET", **kwargs) -> requests.Response:
    """
    Make an API request with x402 payment handling
    
    This function:
    1. Makes initial request
    2. If 402 returned, creates payment signature
    3. Retries with payment header
    """
    url = f"{API_BASE}{endpoint}"
    
    # Initial request
    response = requests.request(method, url, **kwargs)
    
    # Handle 402 Payment Required
    if response.status_code == 402:
        print(f"💰 Payment required for {endpoint}")
        
        data = response.json()
        requirements = parse_402_response(data)
        
        if requirements:
            print(f"  Network: {requirements.network}")
            print(f"  Amount: {requirements.amount} (USDC units)")
            print(f"  PayTo: {requirements.pay_to}")
            
            # Create payment signature
            signature = create_payment_signature(requirements)
            
            # Retry with payment header
            headers = kwargs.get("headers", {})
            headers["X-Payment"] = signature
            
            response = requests.request(method, url, headers=headers, **kwargs)
            
            if response.ok:
                print(f"  ✅ Payment successful!")
    
    return response


# =============================================================================
# API FUNCTIONS
# =============================================================================

def fetch_news() -> Dict[str, Any]:
    """Fetch crypto news with payment"""
    print("\n📰 Fetching crypto news...")
    
    response = make_paid_request("/api/v1/news")
    
    if response.ok:
        data = response.json()
        articles = data.get("articles", [])
        print(f"✅ Received {len(articles)} articles")
        if articles:
            print(f"  Latest: {articles[0].get('title', 'N/A')}")
        return data
    else:
        print(f"❌ Error: {response.status_code}")
        return {}


def fetch_trending() -> Dict[str, Any]:
    """Fetch trending coins"""
    print("\n📈 Fetching trending coins...")
    
    response = make_paid_request("/api/v1/trending")
    
    if response.ok:
        data = response.json()
        coins = data.get("coins", [])
        print(f"✅ Received {len(coins)} trending coins")
        return data
    else:
        print(f"❌ Error: {response.status_code}")
        return {}


def fetch_market_data() -> Dict[str, Any]:
    """Fetch market overview"""
    print("\n📊 Fetching market data...")
    
    response = make_paid_request("/api/v1/market-data")
    
    if response.ok:
        data = response.json()
        print(f"✅ Market cap: ${data.get('total_market_cap', 0):,.0f}")
        return data
    else:
        print(f"❌ Error: {response.status_code}")
        return {}


def check_api_discovery() -> Dict[str, Any]:
    """Check available endpoints via Bazaar discovery"""
    print("\n🔍 Checking available endpoints...")
    
    response = requests.get(f"{API_BASE}/api/.well-known/x402")
    
    if response.ok:
        data = response.json()
        resources = data.get("resources", [])
        print(f"✅ Found {len(resources)} paid endpoints:")
        for r in resources[:5]:
            print(f"  - {r['path']} ({r['price']})")
        if len(resources) > 5:
            print(f"  ... and {len(resources) - 5} more")
        return data
    else:
        print(f"❌ Discovery failed: {response.status_code}")
        return {}


# =============================================================================
# MAIN
# =============================================================================

def main():
    print("🚀 x402 Payment Client - Python Example")
    print("=" * 50)
    
    # Check discovery endpoint first (free)
    check_api_discovery()
    
    # Fetch data with payments
    fetch_news()
    fetch_trending()
    fetch_market_data()
    
    print("\n" + "=" * 50)
    print("✅ Examples complete!")
    print("\nNote: This is a demo. For production, use proper x402 SDK:")
    print("  - pip install x402-client (when available)")
    print("  - Or use the @x402/fetch npm package with Node.js")


if __name__ == "__main__":
    main()

File: market-data/crypto-news/examples/x402-client.go
// x402 Payment Client Example - Go
//
// This example shows how to make authenticated API calls using x402 payments.
//
// Requirements:
//   go get github.com/ethereum/go-ethereum
//
// Usage:
//   export WALLET_PRIVATE_KEY="your_private_key_here"
//   go run x402-client.go
//
// See: https://docs.x402.org

package main

import (
	"bytes"
	"crypto/ecdsa"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/ethereum/go-ethereum/common/hexutil"
	"github.com/ethereum/go-ethereum/crypto"
)

// =============================================================================
// CONFIGURATION
// =============================================================================

const APIBase = "https://cryptocurrency.cv"

var privateKey *ecdsa.PrivateKey
var walletAddress string

func init() {
	keyHex := os.Getenv("WALLET_PRIVATE_KEY")
	if keyHex == "" {
		fmt.Println("Warning: WALLET_PRIVATE_KEY not set. Using demo mode.")
		keyHex = "0x" + strings.Repeat("0", 64)
	}

	// Remove 0x prefix if present
	keyHex = strings.TrimPrefix(keyHex, "0x")

	keyBytes, err := hex.DecodeString(keyHex)
	if err != nil {
		fmt.Printf("Error decoding private key: %v\n", err)
		os.Exit(1)
	}

	privateKey, err = crypto.ToECDSA(keyBytes)
	if err != nil {
		fmt.Printf("Error creating private key: %v\n", err)
		os.Exit(1)
	}

	publicKey := privateKey.Public()
	publicKeyECDSA, ok := publicKey.(*ecdsa.PublicKey)
	if !ok {
		fmt.Println("Error casting public key")
		os.Exit(1)
	}

	walletAddress = crypto.PubkeyToAddress(*publicKeyECDSA).Hex()
	fmt.Printf("Wallet address: %s\n", walletAddress)
}

// =============================================================================
// X402 PAYMENT HANDLING
// =============================================================================

// PaymentRequirements represents parsed 402 response
type PaymentRequirements struct {
	Version  int    `json:"x402Version"`
	Network  string `json:"network"`
	Amount   string `json:"maxAmountRequired"`
	PayTo    string `json:"payTo"`
	Asset    string `json:"asset"`
	Nonce    string `json:"paymentNonce"`
	Resource string `json:"resource"`
}

// X402Response represents the 402 response structure
type X402Response struct {
	Error   string `json:"error"`
	X402    struct {
		Version int `json:"x402Version"`
		Accepts []struct {
			Network   string `json:"network"`
			Amount    string `json:"maxAmountRequired"`
			PayTo     string `json:"payTo"`
			Asset     string `json:"asset"`
			Nonce     string `json:"paymentNonce"`
			Resource  string `json:"resource"`
		} `json:"accepts"`
	} `json:"x402"`
}

// Parse402Response extracts payment requirements from 402 response
func Parse402Response(body []byte) (*PaymentRequirements, error) {
	var resp X402Response
	if err := json.Unmarshal(body, &resp); err != nil {
		return nil, err
	}

	if len(resp.X402.Accepts) == 0 {
		return nil, fmt.Errorf("no accepted payment methods")
	}

	accept := resp.X402.Accepts[0]
	return &PaymentRequirements{
		Version:  resp.X402.Version,
		Network:  accept.Network,
		Amount:   accept.Amount,
		PayTo:    accept.PayTo,
		Asset:    accept.Asset,
		Nonce:    accept.Nonce,
		Resource: accept.Resource,
	}, nil
}

// CreatePaymentSignature creates an x402 payment signature
//
// Note: This is simplified. Production should use official x402-go SDK.
func CreatePaymentSignature(req *PaymentRequirements) (string, error) {
	// Create payment message
	message := map[string]interface{}{
		"network":   req.Network,
		"amount":    req.Amount,
		"payTo":     req.PayTo,
		"asset":     req.Asset,
		"nonce":     req.Nonce,
		"resource":  req.Resource,
		"timestamp": time.Now().Unix(),
	}

	msgBytes, err := json.Marshal(message)
	if err != nil {
		return "", err
	}

	// Create Ethereum signed message hash
	prefixedMsg := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(msgBytes), msgBytes)
	hash := crypto.Keccak256Hash([]byte(prefixedMsg))

	// Sign the message
	signature, err := crypto.Sign(hash.Bytes(), privateKey)
	if err != nil {
		return "", err
	}

	return hexutil.Encode(signature), nil
}

// MakePaidRequest makes an API request with x402 payment handling
func MakePaidRequest(endpoint, method string, body []byte) ([]byte, error) {
	url := APIBase + endpoint

	// Initial request
	var req *http.Request
	var err error

	if body != nil {
		req, err = http.NewRequest(method, url, bytes.NewReader(body))
	} else {
		req, err = http.NewRequest(method, url, nil)
	}
	if err != nil {
		return nil, err
	}

	client := &http.Client{Timeout: 30 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	// Handle 402 Payment Required
	if resp.StatusCode == 402 {
		fmt.Printf("💰 Payment required for %s\n", endpoint)

		requirements, err := Parse402Response(respBody)
		if err != nil {
			return nil, fmt.Errorf("failed to parse 402 response: %w", err)
		}

		fmt.Printf("  Network: %s\n", requirements.Network)
		fmt.Printf("  Amount: %s (USDC units)\n", requirements.Amount)
		fmt.Printf("  PayTo: %s\n", requirements.PayTo)

		// Create payment signature
		signature, err := CreatePaymentSignature(requirements)
		if err != nil {
			return nil, fmt.Errorf("failed to create signature: %w", err)
		}

		// Retry with payment header
		if body != nil {
			req, _ = http.NewRequest(method, url, bytes.NewReader(body))
		} else {
			req, _ = http.NewRequest(method, url, nil)
		}
		req.Header.Set("X-Payment", signature)

		resp, err = client.Do(req)
		if err != nil {
			return nil, err
		}
		defer resp.Body.Close()

		respBody, err = io.ReadAll(resp.Body)
		if err != nil {
			return nil, err
		}

		if resp.StatusCode == 200 {
			fmt.Println("  ✅ Payment successful!")
		}
	}

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("API error: %d", resp.StatusCode)
	}

	return respBody, nil
}

// =============================================================================
// API FUNCTIONS
// =============================================================================

// NewsResponse represents the news API response
type NewsResponse struct {
	Articles []struct {
		Title  string `json:"title"`
		Source string `json:"source"`
	} `json:"articles"`
}

func fetchNews() {
	fmt.Println("\n📰 Fetching crypto news...")

	data, err := MakePaidRequest("/api/v1/news", "GET", nil)
	if err != nil {
		fmt.Printf("❌ Error: %v\n", err)
		return
	}

	var resp NewsResponse
	if err := json.Unmarshal(data, &resp); err != nil {
		fmt.Printf("❌ Parse error: %v\n", err)
		return
	}

	fmt.Printf("✅ Received %d articles\n", len(resp.Articles))
	if len(resp.Articles) > 0 {
		fmt.Printf("  Latest: %s\n", resp.Articles[0].Title)
	}
}

func checkDiscovery() {
	fmt.Println("\n🔍 Checking available endpoints...")

	resp, err := http.Get(APIBase + "/api/.well-known/x402")
	if err != nil {
		fmt.Printf("❌ Error: %v\n", err)
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var discovery struct {
		Resources []struct {
			Path  string `json:"path"`
			Price string `json:"price"`
		} `json:"resources"`
	}

	if err := json.Unmarshal(body, &discovery); err != nil {
		fmt.Printf("❌ Parse error: %v\n", err)
		return
	}

	fmt.Printf("✅ Found %d paid endpoints:\n", len(discovery.Resources))
	for i, r := range discovery.Resources {
		if i >= 5 {
			fmt.Printf("  ... and %d more\n", len(discovery.Resources)-5)
			break
		}
		fmt.Printf("  - %s (%s)\n", r.Path, r.Price)
	}
}

// =============================================================================
// MAIN
// =============================================================================

func main() {
	fmt.Println("🚀 x402 Payment Client - Go Example")
	fmt.Println(strings.Repeat("=", 50))

	// Check discovery endpoint first (free)
	checkDiscovery()

	// Fetch data with payments
	fetchNews()

	fmt.Println()
	fmt.Println(strings.Repeat("=", 50))
	fmt.Println("✅ Examples complete!")
	fmt.Println()
	fmt.Println("Note: This is a demo. For production, use:")
	fmt.Println("  go get github.com/coinbase/x402-go")
}
